hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/TokenTupleUtils.java

```diff
@@ -233,17 +236,30 @@ public static Tuple nftTokenInfoTupleFor(
             @NonNull final Nft nft,
             final long serialNumber,
             @NonNull final String ledgerId,
-            @NonNull final Account ownerAccount) {
-        requireNonNull(ownerAccount);
-        final var nftMetaData = nft.metadata() != null ? nft.metadata().toByteArray() : Bytes.EMPTY.toByteArray();
+            @NonNull final HederaNativeOperations nativeOperations) {
+        requireNonNull(nft);
+        requireNonNull(token);
+        requireNonNull(ledgerId);
+        requireNonNull(nativeOperations);
 
+        final var nftMetaData = nft.metadata() != null ? nft.metadata().toByteArray() : Bytes.EMPTY.toByteArray();
         return Tuple.of(
                 tokenInfoTupleFor(token, ledgerId),
                 serialNumber,
-                headlongAddressOf(ownerAccount),
+                // The odd construct allowing a token to not have a treasury account set is to accommodate
+                // Token.DEFAULT being passed into this method, which a few HtsCall implementations do
+                priorityAddressOf(nft.ownerIdOrElse(token.treasuryAccountIdOrElse(ZERO_ACCOUNT_ID)), nativeOperations),
                 nft.mintTimeOrElse(new Timestamp(0, 0)).seconds(),
                 nftMetaData,
-                headlongAddressOf(nft.spenderIdOrElse(ZERO_ACCOUNT_ID)));
+                priorityAddressOf(nft.spenderIdOrElse(ZERO_ACCOUNT_ID), nativeOperations));
+    }
+
+    private static Address priorityAddressOf(
```

Comment:
Haven't seen the term "priority address" before.  It's in the title of this PR but otherwise not.  What does it mean?

I do see a comment way over at `EvmFrameState` which is not an obvious spot to look. Maybe the comment here?  (And based on that comment I'm still not sure why it is a "priority" address.) (We really do need a glossary of addresses ...)

Reply:
Any time a contract has an `0xab...cd` EVM address (outside the long-zero subspace), it _must_ be referenced by that address anywhere inside the EVM. (This is to avoid creating issues with contracts that use addresses as e.g. keys in a ERC-20 `balances` map; or store a particular address as a privileged account.) 

I.e. `0xab...cd` addresses take "priority" over long-zero addresses.